pkg/image/docker/manifest.go

```diff
@@ -1,38 +1,147 @@
 package docker
 
 import (
+	"bytes"
 	"encoding/json"
 	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
 
+	"github.com/anchore/stereoscope/internal/log"
+	"github.com/anchore/stereoscope/pkg/file"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
 	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/google/go-containerregistry/pkg/v1/types"
 )
 
-type manifest struct {
-	raw    []byte
+type dockerManifest struct {
 	parsed tarball.Manifest
 }
 
 // newManifest creates a new manifest struct from the given Docker archive manifest bytes
-func newManifest(raw []byte) (manifest, error) {
+func newManifest(raw []byte) (*dockerManifest, error) {
 	var parsed tarball.Manifest
 	if err := json.Unmarshal(raw, &parsed); err != nil {
-		return manifest{}, fmt.Errorf("unable to parse manifest.json: %w", err)
+		return nil, fmt.Errorf("unable to parse manifest.json: %w", err)
 	}
 
 	if len(parsed) == 0 {
-		return manifest{}, fmt.Errorf("no valid manifest.json found")
+		return nil, fmt.Errorf("no valid manifest.json found")
 	}
 
-	return manifest{
-		raw:    raw,
+	return &dockerManifest{
 		parsed: parsed,
 	}, nil
 }
 
-// tags returns the image tags referenced within the images manifest file (within the given docker image tar).
-func (m manifest) tags() (tags []string) {
+// allTags returns the image tags referenced within the images manifest file (within the given docker image tar).
+func (m dockerManifest) allTags() (tags []string) {
 	for _, entry := range m.parsed {
 		tags = append(tags, entry.RepoTags...)
 	}
 	return tags
 }
+
+// extractManifest is helper function for extracting and parsing a docker image manifest (V2) from a docker image tar.
+func extractManifest(tarPath string) (*dockerManifest, error) {
+	f, err := os.Open(tarPath)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		err := f.Close()
+		if err != nil {
+			log.Errorf("unable to close tar file (%s): %w", f.Name(), err)
+		}
+	}()
+
+	manifestReader, err := file.ReaderFromTar(f, "manifest.json")
+	if err != nil {
+		return nil, err
+	}
+
+	contents, err := ioutil.ReadAll(manifestReader)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read manifest.json: %w", err)
+	}
+	return newManifest(contents)
+}
+
+// generateOCIManifest takes a docker manifest and a path to the tar and generates an OCI manifest derived from the given arguments and the docker config.
+func generateOCIManifest(tarPath string, manifest *dockerManifest) (*v1.Manifest, error) {
+	f, err := os.Open(tarPath)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		err := f.Close()
+		if err != nil {
+			log.Errorf("unable to close tar file (%s): %w", f.Name(), err)
```

Comment:
This might not be helpful here, but it triggered a thought — for handling errors in deferred functionality, lately I've been seeing people handle this by using named return parameters (having the `error` return value be named `err` from the start of the function's execution, and just setting that value if need be from within the deferred block) — and then logging, too, if desired.

In this specific case, I could see that a problem encountered while closing the TAR file might not be actionable by the consumer, so this is probably good as is 👌

Reply:
interesting point, I hadn't considered that option; re: deferred errors I'll give that some thought